server/src/main/java/org/opensearch/index/shard/FileUploader.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.shard;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.opensearch.common.CheckedFunction;
+import org.opensearch.index.store.RemoteSegmentStoreDirectory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This class is a wrapper over the copying of file from local to remote store allowing to decorate the actual copy
+ * method along with adding hooks of code that can be run before, on success and on failure.
+ *
+ * @opensearch.internal
+ */
+public class FileUploader {
```

Comment:
Can FileUploader leverage or converge with TransferManager

Reply:
This class is not named correctly. This isn't actually uploading any files but wrapping the upload. Is there a need of this class? @raghuvanshraj ?

Reply:
@Bukhtawar we can certainly combine the logic here with S3TransferManager since effectively we are just registering trackers/listeners for different parts of the upload. As of now, we have kept things simple by merging our changes with changes made by @ashking94 for enabling stats tracking for S3 uploads.

@vikasvb90 I kept the name of this class same as what @ashking94 had named it. While it is not performing the upload itself, it is facilitating it by exposing wrappers around the methods used for actual uploads. That being said, we can think of a better name here.